File: SORTING/MergeSort/MergeSort.java
package SORTING.MergeSort;

import java.util.Arrays;

/**
 * MergeSort
 */
public class MergeSort {

    public static void main(String[] args) {
        int[] arr = { 5, 4, 3, 2, 1 };
        System.out.println("Before sorting: " + Arrays.toString(arr));
        System.out.println("After sorting: " + Arrays.toString(mergeSort(arr)));

    }

    // merge sort
    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) {
            return arr;
        }
        int mid = arr.length / 2;
        int[] left = Arrays.copyOfRange(arr, 0, mid);
        int[] right = Arrays.copyOfRange(arr, mid, arr.length);
        return merge(mergeSort(left), mergeSort(right));
    }

    static int[] merge(int[] mergeSort, int[] mergeSort2) {
        int[] result = new int[mergeSort.length + mergeSort2.length];
        int i = 0, j = 0, k = 0;
        while (i < mergeSort.length && j < mergeSort2.length) {
            if (mergeSort[i] < mergeSort2[j]) {
                result[k] = mergeSort[i];
                i++;
            } else {
                result[k] = mergeSort2[j];
                j++;
            }
            k++;
        }
        while (i < mergeSort.length) {
            result[k] = mergeSort[i];
            i++;
            k++;
        }
        while (j < mergeSort2.length) {
            result[k] = mergeSort2[j];
            j++;
            k++;
        }
        return result;
    }
}

File: Recursion/JS/sort/mergesort.js
var nums = [5, 4, 3, 2, 1];
var mergeSort = (nums) => {
    if (nums.length <= 1) {
        return nums;
    }
    var mid = Math.floor(nums.length / 2);
    var left = mergeSort(nums.slice(0, mid));
    var right = mergeSort(nums.slice(mid));
    console.log(left, right);
    return merge(left, right);
}
var merge = (left, right) => {
    var result = [];
    var i = 0;
    var j = 0;
    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }
    while (i < left.length) {
        result.push(left[i]);
        i++;
    }
    while (j < right.length) {
        result.push(right[j]);
        j++;
    }
    return result;
}
console.log(mergeSort(nums));
var obj = {
    name: [1, 12, 12, 12, 12],
    age: [1, 2, 3, 4, 5]
}
console.log(obj);


